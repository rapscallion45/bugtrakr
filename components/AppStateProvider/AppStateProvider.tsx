import { FC } from 'react';
import Preloader from '../Preloader/Preloader';
import useAppStateProviderController from './AppStateProviderController';

interface AppStateProviderProps {
  children?: any;
}

const AppStateProvider: FC<AppStateProviderProps> = function AppStateProvider({ children = null }) {
  const { loading } = useAppStateProviderController();

  return loading ? <Preloader /> : children;
};
export default AppStateProvider;
