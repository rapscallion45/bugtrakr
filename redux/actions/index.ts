export { default as alertActions } from './alert.actions';
export { default as accountActions } from './account.actions';
