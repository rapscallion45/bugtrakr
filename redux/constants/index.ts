export { default as alertConstants } from './alert.constants';
export { default as accountConstants } from './account.constants';
