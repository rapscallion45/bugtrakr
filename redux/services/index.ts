export { default as accountService } from './account.service';
export { default as projectsService } from './projects.service';
