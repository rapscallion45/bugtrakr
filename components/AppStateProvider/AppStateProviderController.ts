import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { accountActions } from '../../redux/actions';

const AppStateProviderController = function AppStateProviderController() {
  const dispatch = useDispatch();
  const { authenticating, user, loggedIn } = useSelector((state) => state.authentication);
  const loading = Boolean(authenticating);

  useEffect(() => {
    /* On first load run user validation */
    dispatch(accountActions.authenticate());
  }, []);

  useEffect(() => {
    /* whenever authentication state changes, grab account info */
    dispatch(accountActions.get(user?.id));
  }, [user, loggedIn]);

  return { loading };
};
export default AppStateProviderController;
