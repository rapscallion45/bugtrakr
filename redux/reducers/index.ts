import { combineReducers } from 'redux';

import registration from './registration.reducer';
import verifyEmail from './verifyEmail.reducer';
import changePassword from './changePassword.reducer';
import resetPassword from './resetPassword.reducer';
import validateResetToken from './validateResetToken.reducer';
import account from './account.reducer';
import alert from './alert.reducer';
import projects from './projects.reducer';
import bugs from './bugs.reducer';
import users from './users.reducer';

const rootReducer = combineReducers({
  account,
  registration,
  verifyEmail,
  changePassword,
  resetPassword,
  validateResetToken,
  alert,
  projects,
  bugs,
  users,
});
export default rootReducer;

export type AppState = ReturnType<typeof rootReducer>;
