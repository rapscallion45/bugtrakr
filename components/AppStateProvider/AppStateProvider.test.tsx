import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { waitFor } from '@testing-library/react';
import rootReducer from '../../redux/reducers';
import AppStateProvider from './AppStateProvider';

const middleware = [thunkMiddleware];
const TEST_DIV_TEXT = "I'm the test child div!";
const MockChild = function MockChild() {
  return <div className="test-div">{TEST_DIV_TEXT}</div>;
};

describe('App State Provider', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Renders correctly enzyme', () => {
    /* Arrange */
    const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));

    /* Act */
    const view = render(
      <Provider store={testStore}>
        <AppStateProvider>
          <MockChild />
        </AppStateProvider>
      </Provider>
    );

    /* Assert */
    expect(toJson(view)).toMatchSnapshot();
  });

  describe('Call relevant API', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call 2 API in total', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));

      /* Act */
      mount(
        <Provider store={testStore}>
          <AppStateProvider>
            <MockChild />
          </AppStateProvider>
        </Provider>
      );

      /* Assert */
      expect(window.fetch).toHaveBeenCalledTimes(2);
    });

    it('Should call Account data API', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));

      /* Act */
      mount(
        <Provider store={testStore}>
          <AppStateProvider>
            <MockChild />
          </AppStateProvider>
        </Provider>
      );

      /* Assert */
      expect(window.fetch).toHaveBeenCalledWith(
        '/api/account',
        expect.objectContaining({
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
        })
      );
    });

    it('Should call authenticate user API', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));

      /* Act */
      mount(
        <Provider store={testStore}>
          <AppStateProvider>
            <MockChild />
          </AppStateProvider>
        </Provider>
      );

      /* Assert */
      expect(window.fetch).toHaveBeenCalledWith(
        '/api/authenticate',
        expect.objectContaining({
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
        })
      );
    });
  });

  describe('Show/Hide Preloader', () => {
    it('Should render Preloader intially when app data not loaded', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <AppStateProvider>
            <MockChild />
          </AppStateProvider>
        </Provider>
      );

      /* Assert */
      expect(view.find('Preloader').length).toBe(1);
      expect(view.find('.test-div').length).toBe(0);
    });

    it('Should render children components once app data loaded', async () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <AppStateProvider>
            <MockChild />
          </AppStateProvider>
        </Provider>
      );

      /* Assert */
      expect(view.find('Preloader').length).toBe(1);
      expect(view.find('.test-div').length).toBe(0);

      await waitFor(() => {
        /* After API calls have completed, preloader no longer displayed */
        view.update();
        expect(view.find('Preloader').length).toBe(0);
      });
      expect(view.find('.test-div').length).toBe(1);
    });
  });
});
