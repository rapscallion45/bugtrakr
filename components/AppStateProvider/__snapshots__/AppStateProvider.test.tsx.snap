// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App State Provider Renders correctly enzyme 1`] = `
<div
  class="MuiBox-root css-c1a2tc"
>
  <div
    class="MuiBox-root css-1tqvppq"
  >
    <img
      class="MuiBox-root css-1domaf0"
      src="/static/logo.png"
    />
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-hbct9n-MuiCircularProgress-root"
      role="progressbar"
      style="width:40px;height:40px"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</div>
`;
